ipc-queue/src/sealed.rs

```diff
@@ -0,0 +1,103 @@
+use std::sync::atomic::{AtomicU64, Ordering};
+
+#[cfg(target_env = "sgx")]
+use std::os::fortanix_sgx::usercalls::alloc::UserSafeSized;
+
+#[cfg(not(target_env = "sgx"))]
+pub trait UserSafeSized: Copy + Sized {}
+
+#[cfg(not(target_env = "sgx"))]
+impl<T> UserSafeSized for T where T: Copy + Sized {}
+
+#[repr(C)]
+pub struct WithId<T: Copy> {
```

Comment:
Actually I meant for this to go in `fortanix-sgx-abi`. Then you can also get rid of most of this module.

Reply:
I'm not sure what you intend here, can you elaborate more? Do you just want the `WithId` and `Identified` to be in the abi crate? or do you want a different restructuring?

Reply:
`WithId` in the abi crate, I don't think you need the `Identified` trait.

Reply:
What bound do you put on T for Sender/Receiver types then?

Reply:
`Copy` (or `UserSafeSized` in the enclave)

```
pub(crate) struct FifoInner<T> {
    data: NonNull<[UnsafeCell<WithId<T>>]>,
    offsets: NonNull<AtomicUsize>,
    storage: Storage<T>,
}

```

Reply:
This would mean that send/recv functions would be changed to take/return id explicitly. Also ultimately we need to transmute `FifoDescriptor<Usercall>` to `FifoDescriptor<WithId<(u64, u64, u64, u64, u64)>>` (or the reverse) somewhere. That does not seem like a better interface to me.